core/src/main/scala/kafka/log/LogValidator.scala

```diff
@@ -78,6 +78,12 @@ private[kafka] object LogValidator extends Logging {
       if (batch.isControlBatch)
         throw new InvalidRecordException("Clients are not allowed to write control records")
     }
+
+    if (batch.isTransactional && toMagic < RecordBatch.MAGIC_VALUE_V2)
+      throw new UnsupportedForMessageFormatException(s"Transactional records cannot be used with magic version $toMagic")
```

Comment:
Is this ever expected to happen? If it is then we should not throw an exception to fail the broker but return an error code back to client.